### PR TITLE
Add distribute strategies to transformer.

### PR DESCRIPTION
**Pull request by isaprykin**

### CODE REVIEW

File: official/transformer/v2/metrics.py [hunk 1]
```diff
@@ -160,9 +160,10 @@ def get_config(self):
 
   def call(self, inputs):
     logits, targets = inputs[0], inputs[1]
-    for mean, fn in self.metric_mean_fns:
-      m = mean(*fn(logits, targets))
-      self.add_metric(m)
+    if not tf.distribute.in_cross_replica_context():
```
Comment (tfboyd) [inline official/transformer/v2/metrics.py]: Leave a comment and bug link if relevant as to why this is needed, it seems odd that we have to do something special.

File: official/transformer/v2/transformer.py [hunk 2]
```diff
@@ -103,6 +103,9 @@ def call(self, x, training):
           outputs: [batch_size, decoded length]
           scores: [batch_size, float]}
     """
+    if not tf.distribute.in_cross_replica_context() and len(x) == 1:
```
Comment (tfboyd) [inline official/transformer/v2/transformer.py]: Leave a comment and bug link if relevant as to why this is needed, it seems odd that we have to do something special.

File: official/transformer/v2/transformer_benchmark.py [hunk 3]
```diff
@@ -142,6 +142,7 @@ def benchmark_1_gpu(self):
     FLAGS.batch_size = 4096
     FLAGS.train_steps = 200000
     FLAGS.steps_between_evals = 5000
+    FLAGS.static_batch = True
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: Make a new test called   `benchmark_1_gpu_static`  so we can track the difference.  legacy graph does not seem to have this issue or not as bad running the same keras code.  

Once they both converge or we are sure of the standard we can drop one of them.  I am not certain at this time what is "right".  I don't want to run both to convergence, but for now both seems correct.

File: official/transformer/v2/transformer_benchmark.py [hunk 4]
```diff
@@ -150,6 +151,54 @@ def benchmark_1_gpu(self):
...
+    """
+    self._setup()
+    FLAGS.num_gpus = 8
+    FLAGS.data_dir = self.train_data_dir
+    FLAGS.vocab_file = self.vocab_file
+    # Sets values directly to avoid validation check.
+    FLAGS['bleu_source'].value = self.bleu_source
+    FLAGS['bleu_ref'].value = self.bleu_ref
+    FLAGS.param_set = 'base'
+    FLAGS.batch_size = 32768
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: Let's write this out as 8 * 4096 to make the per gpu batch-size obvious at a glance.  And change the steps to 100K.  Running 100k I reached 27 to 27.2 but never then 27.3 desired.

File: official/transformer/v2/transformer_benchmark.py [hunk 5]
```diff
@@ -150,6 +151,54 @@ def benchmark_1_gpu(self):
                                    bleu_min=25.3,
                                    bleu_max=26)
 
+  def benchmark_8_gpu(self):
+    """Benchmark 8 gpu.
+
+      The paper uses 8 GPUs and a much larger effective batch size, this is will
+      not converge to the 27.3 BLEU (uncased) SOTA.
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: Change this to say:
"Should converge to 27.3 BLEU (uncased). This has not been confirmed yet."

File: official/transformer/v2/transformer_benchmark.py [hunk 6]
```diff
@@ -150,6 +151,54 @@ def benchmark_1_gpu(self):
...
+    FLAGS.train_steps = 200000
+    FLAGS.static_batch = True
+    FLAGS.steps_between_evals = 5000
+    FLAGS.model_dir = self._get_model_dir('benchmark_8_gpu')
+    self._run_and_report_benchmark(total_batch_size=FLAGS.batch_size,
+                                   log_steps=FLAGS.log_steps,
+                                   bleu_min=25.3,
+                                   bleu_max=26)
+
+  def benchmark_big_8_gpu(self):
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: Make a new class: TransformerBigKerasAccuracy and then just use `benchmark_8_gpu`

I look for the word  `TransformerBig` as a way to know the model when I do reporting.

File: official/transformer/v2/transformer_benchmark.py [hunk 7]
```diff
@@ -150,6 +151,54 @@ def benchmark_1_gpu(self):
...
+    """Benchmark 8 gpu.
+
+      The paper uses 8 GPUs and a much larger effective batch size, this is will
+      not converge to the 27.3 BLEU (uncased) SOTA.
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: Change this to say (you could add a date, but I am pretty sure we will remember to update when it converged like I did with the estimator version:

"Should converge to 28.4 BLEU (uncased). This has not be verified yet."

File: official/transformer/v2/transformer_benchmark.py [hunk 8]
```diff
@@ -150,6 +151,54 @@ def benchmark_1_gpu(self):
...
+    self._setup()
+    FLAGS.num_gpus = 8
+    FLAGS.data_dir = self.train_data_dir
+    FLAGS.vocab_file = self.vocab_file
+    # Sets values directly to avoid validation check.
+    FLAGS['bleu_source'].value = self.bleu_source
+    FLAGS['bleu_ref'].value = self.bleu_ref
+    FLAGS.param_set = 'big'
+    FLAGS.batch_size = 24576
+    FLAGS.train_steps = 200000
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: I think 100,000 steps is enough.  But we have not validated yet.  Totally FYI not a big deal until we know.

File: official/transformer/v2/transformer_benchmark.py [hunk 9]
```diff
@@ -150,6 +151,54 @@ def benchmark_1_gpu(self):
...
+    FLAGS.static_batch = True
+    FLAGS.steps_between_evals = 5000
+    FLAGS.model_dir = self._get_model_dir('benchmark_big_8_gpu')
+    self._run_and_report_benchmark(total_batch_size=FLAGS.batch_size,
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: This should be something like 28.0 to 29.0 for bleu scores, we will need to dial it in.  Ideally we would have the min be 28.4 matching the paper. More of an FYI we can dial this in later.

File: official/transformer/v2/transformer_benchmark.py [hunk 10]
```diff
@@ -179,6 +228,7 @@ def benchmark_1_gpu(self):
     FLAGS.num_gpus = 1
     FLAGS.batch_size = self.batch_per_gpu
     FLAGS.model_dir = self._get_model_dir('benchmark_1_gpu')
+    FLAGS.static_batch = True
```
Comment (tfboyd) [inline official/transformer/v2/transformer_benchmark.py]: I would rather a new test with `benchmark_1_gpu_static`

File: official/transformer/v2/transformer_main.py [hunk 11]
```diff
@@ -107,13 +113,22 @@ def train(self):
...
+    else:
+      model = transformer.create_model(params, is_train)
+      opt = self._create_optimizer()
+      model.compile(opt)
 
-    model.compile(opt, target_tensors=[])
     model.summary()
 
-    map_data_fn = data_pipeline.map_data_for_transformer_fn
+    if self.distribution_strategy:
```
Comment (tfboyd) [inline official/transformer/v2/transformer_main.py]: Might be nice to have a comment as to why you do this for dist_strat.  This is kind of a learning model and nice to point people to it and they can understand the why.

File: official/transformer/v2/transformer.py (hunk 2 above)
Comment (guptapriya) [reply official/transformer/v2/transformer.py]: Figured out why this was needed and changed the mode such that this is no longer required.

File: official/transformer/v2/metrics.py (hunk 1 above)
Comment (guptapriya) [reply official/transformer/v2/metrics.py]: Found another way to write metrics that doesn't need this hack. Bug filed to track the issue with the old approach.

File: official/transformer/v2/transformer_main.py (hunk 11 above)
Comment (guptapriya) [reply official/transformer/v2/transformer_main.py]: added a comment. I tried hard to write the input pipeline in a way that works in both cases but failed so far. A proper fix may need a change in keras i think. this should not be the end state..

**Comment by isaprykin** [comment]
I don't get these test failures.  I wonder if they are stale or related.
Are there any unresolved comments on this PR before we merge?

**Pull request merged**